mods/disable-virtual-desktop-transition.wh.cpp

```diff
@@ -0,0 +1,205 @@
+
+class CVersionHelper
+{
+public:
+	// Data structure returned by GetVersionInfo().
+	struct VersionStruct
+	{
+		bool isInitialized = false;
+		DWORD dwMajorVersion = 0;
+		DWORD dwMinorVersion = 0;
+		DWORD dwBuildNumber = 0;
+		DWORD dwPlatformId = 0;
+	};
+
+	typedef void (WINAPI *RtlGetVersion_t)(OSVERSIONINFOEXW *);
+
+	// Gets the precise OS version.
+	static const VersionStruct *GetVersionInfo()
+	{
+		static VersionStruct s_versionStruct = { 0 };
+
+		// Skip if cached.
+		if (!s_versionStruct.isInitialized)
+		{
+			HMODULE hMod = LoadLibraryW(L"ntdll.dll");
+
+			if (hMod)
+			{
+				RtlGetVersion_t func = (RtlGetVersion_t)GetProcAddress(hMod, "RtlGetVersion");
+
+				if (!func)
+				{
+					FreeLibrary(hMod);
+
+					// TODO: error handling.
+					return &s_versionStruct;
+				}
+
+				OSVERSIONINFOEXW osVersionInfo = { 0 };
+				osVersionInfo.dwOSVersionInfoSize = sizeof(OSVERSIONINFOEXW);
+
+				func(&osVersionInfo);
+
+				s_versionStruct.dwBuildNumber = osVersionInfo.dwBuildNumber;
+				s_versionStruct.dwMajorVersion = osVersionInfo.dwMajorVersion;
+				s_versionStruct.dwMinorVersion = osVersionInfo.dwMinorVersion;
+				s_versionStruct.dwPlatformId = osVersionInfo.dwPlatformId;
+
+				s_versionStruct.isInitialized = true;
+
+				FreeLibrary(hMod);
+			}
+		}
+
+		return &s_versionStruct;
+	}
+
+	// Specific version helpers.
+	inline static bool IsWindows10OrGreater()
+	{
+		return GetVersionInfo()->dwMajorVersion >= 10 &&
+			   GetVersionInfo()->dwBuildNumber >= 10240;
+	}
+
+    inline static bool IsWindows11OrGreater()
+    {
+        return GetVersionInfo()->dwMajorVersion >= 10 &&
+               GetVersionInfo()->dwMinorVersion >= 22000;
+    }
+};
+
+__declspec(noinline) HRESULT CVirtualDesktop__Initialize(void *pThis, void *a2)
+{
+    // Failure result makes DWM reject the animation and move on like normal.
+    return E_FAIL;
+}
+
+size_t g_curVftableVirtualDesktopSwitch[18];
+
+typedef void *(__cdecl *CVirtualDesktopSwitch__CVirtualDesktopSwitch_t)(void *pThis);
+CVirtualDesktopSwitch__CVirtualDesktopSwitch_t CVirtualDesktopSwitch__CVirtualDesktopSwitch_orig;
+void *__cdecl CVirtualDesktopSwitch__CVirtualDesktopSwitch_hook(void *pThis)
+{
+    void *pAnim = CVirtualDesktopSwitch__CVirtualDesktopSwitch_orig(pThis);
+
+    /*
+     * CVirtualDesktop inherits the Initialize() method from CStoryboard, so we need to
+     * create our own copy of the virtual function table to replace this with our own
+     * method.
+     */
+
+    memcpy(g_curVftableVirtualDesktopSwitch, *(void **)pAnim, 18 * sizeof(size_t));
+    g_curVftableVirtualDesktopSwitch[1] = (size_t)CVirtualDesktop__Initialize;
+    *(void **)pAnim = (void *)g_curVftableVirtualDesktopSwitch;
+
+    return pAnim;
+}
+
+// The mod is being initialized, load settings, hook functions, and do other
+// initialization stuff if required.
+BOOL Wh_ModInit()
+{
+    bool fAlreadyRunning = !(*(USHORT*)(((long long)NtCurrentTeb()) + 0x17EE) & 0x0400);
```

Comment:
Can't `fAlreadyRunning` be false on system start? I think it's better to make the check stricter, e.g. by checking that it happens more than once in a short interval for the same session.

Reply:
It's never failed to work on system startup for me.

Reply:
^ Or for other users who have tested it. It only cancels initialisation if DWM is restarted due to a crash, although I don't understand how it works, just that it works. If you disable the mod and re-enable it, then Windhawk will also succeed to inject it just fine.

Reply:
Looks like on boot, dwm.exe is launched by wininit.exe which is a protected process. That's why Windhawk isn't able to intercept the process creation, and the injection happens asynchronously. But that doesn't mean that `fAlreadyRunning` will be always true in this case. If Windhawk is quick enough with the asynchronously injection, it will inject its code before dwm.exe actually starts running, and `fAlreadyRunning` will be false. Also, Windhawk might improve in the future to inject synchronously. So while this check is fine for testing, I'd not include it in this form. I suggest to either remove it or add a more robust check.

Reply:
I thought for alternatives, and came up with this function:
https://gist.github.com/m417z/58d2fffd846c358b961f8c70d4f46c0d

Looks like a warning is logged when DWM restarts, it should be a more reliable check.

![image](https://github.com/user-attachments/assets/af17002f-5226-46b3-bea7-b82f153eaf1d)

Reply:
Latest commit resolves this issue.